File: python/simple-linked-list/simple_linked_list.py
class Node(object):
    def __init__(self, value):
        self._value = value
        self._next = None

    def value(self):
        return self._value

    def next(self):
        return self._next

class LinkedList(object):
    def __init__(self, values=[]):
        self._size = 0
        self._head= None
        for value in values:
            self.push(value)

    def __len__(self):
        return self._size

    def head(self):
        if self._size == 0:
            raise EmptyListException('List is empty')
        return self._head

    def push(self, value):
        self._size += 1
        new_node = Node(value)
        new_node._next = self._head
        self._head = new_node

    def pop(self):
        res = self.head().value()
        self._head = self.head().next()
        self._size -= 1
        return res

    def reversed(self):
        return reversed(list(self))

    def __iter__(self):
        self.curr_node = Node(None)
        self.curr_node._next = self._head
        return self
    
    def __next__(self):
        if self.curr_node._next == None:
            raise StopIteration
        self.curr_node = self.curr_node._next
        return self.curr_node._value

class EmptyListException(Exception):
    pass

File: java/simple-linked-list/src/main/java/SimpleLinkedList.java
import java.util.List;
import java.util.ArrayList;
import java.util.NoSuchElementException;

class Node<T> {
  T data;
  Node<T> next;

  Node(T data) {
    this.data = data;
    next = null;
  }
}
@SuppressWarnings("unchecked")
class SimpleLinkedList<T> {
  int size = 0;
  Node<T> head = null;

  SimpleLinkedList() {}

  SimpleLinkedList(T[] values) {
    for (int i = 0; i < values.length; i++) {
      push(values[i]);
    }
  }

  public int size() {
    return size;
  }

  public void push(T value) {
    Node<T> newNode = new Node<>(value);
    Node<T> temp = head;
    if (temp == null) {
      head = newNode;
    } else {
      while (temp != null && temp.next != null) {
        temp = temp.next;
      }
      temp.next = newNode;
    }
    size++;
  }

  public T pop() {
    if (size() == 0) {
      throw new NoSuchElementException("List is empty");
    }

    T res;

    if (size() == 1) {
      res = head.data;
      head = null;
    } else {
      Node<T> temp = head;
      while (temp.next.next != null) {
        temp = temp.next;
      }
      res = temp.next.data;
      temp.next = null;
    }
    size--;
    return res;
  }

  private Node<T> reverse(Node<T> head) {
    if (head == null || head.next == null) {
      return head;
    }
    Node<T> newHead = reverse(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
  }

  public void reverse() {
    if (head == null || head.next == null) {
      return;
    }
    head = reverse(head);
  }

  public T[] asArray(Class<T> c) {
    List<T> res = new ArrayList<T>();
    while (size > 0) {
      res.add(pop());
    }
    return (T[])res.toArray();
  }
}